UrlForm encoding and AWS query client support

dhpiggott opened the pull request:
This is good for review now. Some notes:
1. I still want to run Scalafix against just the 26 paths touched by this PR to ensure imports are organised, but I haven't figured out a good way to do that yet. Not sure if that's essential but definitely nice to have.
2. I haven't waited for CI to be all green since my last change so there could be a surprise failure I've not seen yet. Hopefully not though.
3. Given the number of commits and the presence of decoding stuff in some of the earlier ones, I'd like this to be squash merged.

dhpiggott committed:
UrlForm support

dhpiggott committed:
Wire up codec, client and compliance tests

dhpiggott committed:
Fix tests... except the one that seems like it requires a remodel

dhpiggott committed:
Disgusting hacks

Obviously we don't want this to be the permanent solution but it at
least confirms understanding of the problem.

With this in place all the compliance tests except for the idempotency
token case (which we're ignoring for now anyway) pass.

dhpiggott committed:
Less stinky solve for collection/map encoding difference

Credit to Oli.

dhpiggott committed:
Smoke test the AWS query protocol by listing CloudWatch metrics

dhpiggott committed:
Remove decoding support (for now)

dhpiggott committed:
s/smoking/sandbox/

dhpiggott committed:
Remove commented code

dhpiggott committed:
Disable QueryIdempotencyTokenAutoFill test

dhpiggott committed:
Reenable other compliance tests

dhpiggott committed:
Avoid UrlForm having to be aware of Action and Version hints

Baccata commented on modules/aws-http4s/src/smithy4s/aws/internals/AwsQueryCodecs.scala:
```diff
@@ -0,0 +1,226 @@
+              new PolyFunction[UrlForm.Encoder, EntityEncoder[F, *]] {
+                def apply[A](fa: UrlForm.Encoder[A]): EntityEncoder[F, A] =
+                  urlFormEntityEncoder[F].contramap((a: A) =>
+                    UrlForm(
+                      UrlForm.FormData.MultipleValues(
+                        UrlForm.FormData
+                          .PathedValue(
+                            PayloadPath(PayloadPath.Segment("Action")),
+                            endpoint.id.name
+                          )
+                          .toPathedValues ++
+                          UrlForm.FormData
+                            .PathedValue(
+                              PayloadPath(PayloadPath.Segment("Version")),
+                              version
+                            )
+                            .toPathedValues ++
+                          fa.encode(a).values.values
```
aesthetically speaking, I'm not a fan of [hadouken code](https://i.pinimg.com/originals/98/bd/38/98bd38321fad0e8b14ebf3a5a9ec70c9.jpg) 

I'd rather the `Action`/`Version` related was extracted in a `addOperationHeader` method

dhpiggott replied:
Yeah this mess isn't meant to be final, it's just another step along the way.

dhpiggott replied:
Improved it a bit in https://github.com/disneystreaming/smithy4s/pull/1112/commits/74cf808b3c3a76e7282bde7280bb6e78b5bca5f8. My plan is to add a function like `addValue` in the decoding PR which is the one where I plan to finalise the `UrlForm` model.

dhpiggott committed:
Remove functions not required for encoding

dhpiggott committed:
Remove unused encodesUnion function

dhpiggott committed:
Remove unused EncoderK

dhpiggott committed:
Make better use of UrlFormDataEncoder.contramap and optional

dhpiggott committed:
Inline optional

Because it only had one callsite.

dhpiggott committed:
Remove unused empty values

dhpiggott committed:
Remove redundant sandbox package

dhpiggott committed:
Simplify some syntax

dhpiggott committed:
Fix Scala 2.12

dhpiggott committed:
Remove unused import

dhpiggott committed:
Final refinement of core changes part 1

dhpiggott committed:
Remove unused import

dhpiggott committed:
Final refinement of core changes part 2

dhpiggott committed:
Remove redundant TODO

dhpiggott committed:
Add override keyword

dhpiggott committed:
Remove more commented code

dhpiggott committed:
Be consistent in how URLEncoder.encode calls refer to UTF8

Also remove another unused import...

dhpiggott committed:
Add missing import

dhpiggott committed:
Handle TODO on PathedValue, improve list case readability and comment

dhpiggott committed:
Remove redundant TODO on UrlFormEncoderSchemaVisitorSpec

Because weaver is the preference for tests now, and munit is only still
in use for historical reasons.

dhpiggott committed:
Merge branch 'series/0.18' into url-form-encoding-support

dhpiggott committed:
Make union encoding compile again

dhpiggott committed:
Make Sandbox compile again

NextToken seems to longer be generated - the values are just strings.

dhpiggott committed:
DRY XML response decoding

dhpiggott committed:
Tidy requestEncoderCompilers

It may make sense to add a function on the FormData types that abstracts
the prepending of additional values, but I don't think now is the time
because the FormData model isn't final yet, and won't be until decoding
support is (which is the subject of a separate PR).

dhpiggott committed:
Make UrlForm and encoder schema visitor private[smithy4s]

dhpiggott committed:
Refine Sandbox

dhpiggott committed:
Fix GzipRequestSpec

dhpiggott committed:
Make copyright years pass header check

dhpiggott commented on modules/core/src/smithy4s/http/UrlForm.scala:
```diff
@@ -0,0 +1,146 @@
+
+    }
+
+    // TODO: Rename as Value, replace uses by PathedValue?
```
Will handle this as part of the decoding PR.

dhpiggott commented on modules/core/src/smithy4s/http/UrlForm.scala:
```diff
@@ -0,0 +1,146 @@
+
+    }
+
+    // TODO: Rename as Values?
```
Ditto, because it won't be clear if the model is good until decoding support is finalised.

dhpiggott commented on modules/aws-http4s/src/smithy4s/aws/AwsCredentialsProvider.scala:
```diff
@@ -62,6 +62,8 @@ class AwsCredentialsProvider[F[_]](implicit F: Temporal[F]) {
       .eval(fromEnv)
       .map(F.pure)
+      // TODO: Ensure minimal delay when these endpoints don't exist, e.g.
+      // when running locally.
```
Didn't feel it was appropriate to make any more changes to the provider as part of this PR, so just leaving this here.

Baccata replied:
Fair nuff

dhpiggott committed:
And Sandbox

Baccata commented on modules/sandbox/src/Sandbox.scala:
```diff
@@ -0,0 +1,96 @@
+      cloudWatch <- AwsClient(CloudWatch, awsEnv)
+    } yield cloudWatch
+
+  // This is probably something that's gonna get reimplemented a lot in
+  // user-land. Perhaps we could use pagination hints from the specs to avoid
+  // having to manually wire up the accessors, and to generate synthetic service
+  // functions that handle pagination?
```
yup I have a plan for this

Baccata commented on modules/core/src/smithy4s/http/UrlForm.scala:
```diff
@@ -0,0 +1,146 @@
+
+      override def render: String = {
+        val lastIndex = path.segments.size - 1
+        val renderedKey = path.segments.zipWithIndex
+          .foldLeft(new mutable.StringBuilder) {
+
+            case (builder, (Segment.Label(label), i)) if i < lastIndex =>
+              builder.append(
+                URLEncoder.encode(label, StandardCharsets.UTF_8.name())
+              )
+              builder.append('.')
+
+            case (builder, (Segment.Index(index), i)) if i < lastIndex =>
+              builder.append(index)
+              builder.append('.')
+
+            case (builder, (Segment.Label(label), _)) =>
+              builder.append(
+                URLEncoder.encode(label, StandardCharsets.UTF_8.name())
+              )
+
+            case (builder, (Segment.Index(index), _)) =>
+              builder.append(index)
+          }
+          .toString()
```
I'd rather this was an imperative for-loop, and avoided `zipWithIndex` to save on allocation. Moreover, mixing folds with mutable-data (StringBuilder) is a code-smell.

dhpiggott replied:
I'd left this mostly unchanged relative to the original in https://github.com/disneystreaming/smithy4s/blob/8b8cba7819cef902b3bc81c714f00d431b00125f/modules/aws-http4s/src/smithy4s/aws/query/FormData.scala#L56

Makes sense to improve it like that though! Done in https://github.com/disneystreaming/smithy4s/pull/1112/commits/ff5a7c14dd60fc08bf699551c3a8e77e280a8b02.

Baccata commented on modules/core/src/smithy4s/http/UrlForm.scala:
```diff
@@ -0,0 +1,146 @@
+              builder.append(index)
+          }
+          .toString()
+        val renderedValue = maybeValue match {
+          case None => ""
+          case Some(value) =>
+            URLEncoder.encode(value, StandardCharsets.UTF_8.name())
+        }
+        renderedKey + "=" + renderedValue
```
this could keep appending to the `StringBuilder`, rather than allocating new strings via concatenation

dhpiggott replied:
Ah yeah. Done in https://github.com/disneystreaming/smithy4s/pull/1112/commits/ff5a7c14dd60fc08bf699551c3a8e77e280a8b02.

Baccata commented on modules/core/src/smithy4s/http/UrlForm.scala:
```diff
@@ -0,0 +1,146 @@
+
+  sealed trait FormData extends Product with Serializable {
+    def prepend(segment: PayloadPath.Segment): FormData
+    def render: String
```
I think this should delegate to a `def writeTo(sb: StringBuilder): Unit` method

dhpiggott replied:
Nice. Done in https://github.com/disneystreaming/smithy4s/pull/1112/commits/ff5a7c14dd60fc08bf699551c3a8e77e280a8b02.

dhpiggott committed:
Reduce allocations and impedance mismatch in UrlForm.render

dhpiggott committed:
Further reduce allocations and impedance mismatch in UrlForm.render

dhpiggott committed:
Fix non-Unit value discard warning

dhpiggott committed:
Move render to UrlForm

dhpiggott committed:
Really fix non-Unit value discard warning

Baccata reviewed (approved):
Great work !
